doc/performance-testing/0004_stress-test-sept-2020.md

```diff
@@ -0,0 +1,69 @@
+| 110 |34,000|750|125|0.00007|[output](https://console.cloud.google.com/storage/browser/eq-stress-test-load-injectors-benchmark-outputs/stress-test/2020-09-29T09:14:23)|
+| 120 |37,000|802|115|0.00004|[output](https://console.cloud.google.com/storage/browser/eq-stress-test-load-injectors-benchmark-outputs/stress-test/2020-09-29T09:37:14)|
+
+- 600 Runner instances limit, 1800 (2400) vCPU available (3 of 4 cores)
+- 599 Runner instances reached, 1797 (2396) vCPU requested, 802 vCPU used (45%)
+- Error rate includes any HTTP status codes in the 400 and 500 range (neither are expected)
+- The 99th percentile timings are the load balancer response times reported in Stackdriver
+
+## Errors and Failures
+
+- There were a total of 105 failures as reported by Locust across all requests (232,232,087)
+- The number of Locust failures (105) correlates with the errors we see in the load balancer and application logs and are detailed below
+- 500 x 105 errors reported by load_balancer as `statusDetails: "response_sent_by_backend"`
+- 500 errors x 92 reported by the application logs
+        
+        77: "  File '/usr/local/lib/python3.8/site-packages/google/cloud/datastore/_http.py', line 70, in _request"
+        78: "    response = http.request(url=api_url, method='POST', headers=headers, data=data)"
+        85: "  File '/usr/local/lib/python3.8/site-packages/requests/adapters.py', line 498, in send"
+        86: "    raise ConnectionError(err, request=request)"
+        87: "requests.exceptions.ConnectionError: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))"
+
+- 500 errors x 13 reported by the application logs
+
+        89: "  File '/usr/local/lib/python3.8/site-packages/google/cloud/datastore/_http.py', line 70, in _request"
+        90: "    response = http.request(url=api_url, method='POST', headers=headers, data=data)"
+        97: "  File '/usr/local/lib/python3.8/site-packages/requests/adapters.py', line 498, in send"
+        98: "    raise ConnectionError(err, request=request)"
+        99: "requests.exceptions.ConnectionError: ('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer'))"
+
+## Observations
+
+- The run of 120 load injector instances saw no significant increase in response times
+- At 37k rps Redis CPU reached 91% and allowed Runner to scale to 599 instances (past June scale test of 523 Runner instances). Increasing the Redis instance from 4GB TO 36GB saw ~15% improvement in Redis CPU consumption
+- The average submission rate for our benchmark journey in the 120 instances test was 150 responses per second, which is 540,000 responses per hour (assuming requests remain stable over an hour)
+- 37,000 rps saw 802 used vCPU. This equates to 15.44 rps per core (as a GCP resource where the 4th core is effectively unusable). The test achieved 46 rps per core from the 802 vCPU being utilised at 37,000 rps
```

Comment:
I think we need to choose whether to state numbers against the 3 cores available to runner or the 4 cores on each node, as it's confusing to see the two different comparisons being used throughout. I think it should be against the 3 cores.
```suggestion
- 37,000 rps at 802 used vCPU is 46 rps per core
- 37,000 rps at 1797 requested vCPU is 20.59 rps per core
```

Reply:
The reason I've stated both is that total VM vCPU cores is what CATD have historically reported against (in their original performance testing). This value has been put to me when being asked to update on our performance progress, hence it being included here.

Reply:
The current wording is confusing as it first mentions the used vCPU, then switches to the rps for all cores, then back to the rps for the used vCPU. How about something along the lines of:
```suggestion
- 37,000 rps at 802 used vCPU is 46 rps per core
- 37,000 rps at 2396 vCPU is 15.44 rps per core
```

I also think there is another issue here, as the cluster wasn't pushed to it's limits. Arguably, if we increased the requests further after the cluster vCPU limit was reached, then the rps per cluster core would increase.